### ISSUE TITLE
[feature] add multiple model/optimizer/criterion/scheduler support in single stage

### ISSUE THREAD

**Issue by reyzharkov** [catalyst-team/catalyst#312]
For now there is no support for multiple models/optimizers/etc in the single stage of an experiment, which is essential, for instance, for GAN training.

My suggestion is the following:
1. Refactor `RunnerState` to have all model/optimizer/scheduler/criterior attributes in plural form (model**s**/optimizer**s**/scheduler**s**/criterion**s**) in form of `OrderedDict`. Or alternatively subclass `MultiModelRunnerState(RunnerState)` with such properties.
2. Add `ActivationCallback` - which role is to turn on/off backprop of models and to choose optimizers/schedulers/criterions which are important and needed to be called at the moment (see example below).
3. Support multiple/composite models, multiple optimizers/schedulers/criterions reading from config. Add `activity_params` to optimizers/schedulers/criterions/models in config which specifies when they will be activated.
4. Add simple GAN example which will use all that functionality

The approximate look of the proposed scheme:
```python
# begin batch event
def activity_callback.on_batch_begin(self, state):
        # set models train/eval mode
        if state.need_backward:  # we are in train loader
            for model_name, model in state.models.items():
                active_on, activity_period = state.activity_params[model_name]
                if state.training_step % activity_period in active_on:
                    model.eval()
                else:
                    model.train()
            self.active_optimizers = [is_active(state.training_step, state.activity_params[opt] for opt in state.optimizers]
            self.active_schedulers = [is_active(state.training_step, state.activity_params[opt] for opt in state.schedulers]
            self.active_criterions = [is_active(state.training_step, state.activity_params[opt] for opt in state.criterions]
        else:  # we are in valid/infer loader
            self.active_optimizers = self.optimizers.values()
            self.active_schedulers = self.schedulers.values()
            self.active_criterions = self.criterions.values()

# optimizer_callback sets state.loss = None
# scheduler_callback & criterion_callback do nothing
# begin batch event end

# batch prediction

# end batch event
# activity callback -> do nothing
criterion_callback: for crit in state.active_optimizers: crit.compute_loss()
optimizer_callback: for optim in state.active_optimizers: optim.step()
scheduler_callback: for optim in state.active_optimizers: optim.step()
# end batch event end
```

config.yml (for this part I am unsure what is the best way to write this config to make it look less repetitive):
```yaml
model_params:
  model:
    generator:
      model: SimpleGenerator
      ... # SimpleGenerator params
      activity_period: 6
      active_on: [5]
    discriminator:
      model: SimpleDiscriminator
      ... # SimpleDiscriminator params
      activity_period: 6
      active_on: [0, 1, 2, 3, 4]
```

I think the most complicated/boring part with this may be config organization/parsing/registering models part. The rest is quite clear for me, so the implementation may be devided into:
1. No-config example (pp 1, 2, 4)
2. Config example + parsing + registering (pp 3)

**Comment by reyzharkov** [comment]
implemented in #407
